Fixed #24215 -- Refactor of lazy model operations

AlexHill opened the pull request:
See #3984 for history – this is a new pull request based on master instead of stable/1.8.x.

I've bumped the deprecation warning for ``add_lazy_relation`` to ``RemovedInDjango21``, and documented the deprecation.

I've also consolidated the commits into a smaller number of logical chunks.

CC @timgraham @carljm @charettes @akaariai

AlexHill commented on django/db/migrations/state.py:
```diff
@@ -180,10 +179,10 @@ def __init__(self, real_apps, models, ignore_swappable=False):
+                raise ValueError(msg.format(model=lookup_model))
             else:
-                do_pending_lookups(model)
+                self.do_pending_operations(model)
```
This might be one for @andrewgodwin: this branch is never hit in the test suite. Whenever this for block is entered, ``ValueError`` is raised. The comment says that fields are sometimes added after construction – can we identify when that might happen and add a test, or else rewrite this block to just raise the ``ValueError`` if ``_pending_operations`` isn't empty?

MarkusH replied:
On my pull request #4097 I don't see the `else` executed either in Django's own test suite. The `except`, however, and the `raise ValueError()` inside is actively used. The entire for loop can practically be replaced by the same code which is used during Django's boot process: "here are a bunch of Python modules, go and find all model classes in there and register all relations".

carljm replied:
I'd certainly be tempted to get rid of this branch, unless someone can identify a case that hits it.

AlexHill replied:
I've refactored this section to get rid of it.

MarkusH commented on django/apps/registry.py:
```diff
@@ -329,5 +331,54 @@ def clear_cache(self):
+        the corresponding models have been loaded, call the function with
+        the model classes as its arguments.
+
+        Each element of models_or names may be
```
`models_or_names`

AlexHill replied:
Good catch, thanks

MarkusH commented on django/apps/registry.py:
```diff
@@ -329,5 +331,54 @@ def clear_cache(self):
+
+        # Eagerly parse all model strings so we can fail immediately
+        # if any are plainly invalid.
+        from django.db.models.utils import make_model_tuple
```
Do we need this import here or is it safe to put it at the top of the file? Maybe add a comment if it's a circular import if it needs to stay.

AlexHill replied:
Yeah, it's circular. ``apps`` is imported from ``django.apps.registry`` somewhere in ``django.db.models``. Is an inline comment after the line OK?

MarkusH commented on django/apps/registry.py:
```diff
@@ -208,6 +209,7 @@ def register_model(self, app_label, model):
                     (model_name, app_label, app_models[model_name], model))
         app_models[model_name] = model
         self.clear_cache()
+        self.do_pending_operations(model)
```
I don't understand why you need to resolve all dependencies once you cleared the cache.

AlexHill replied:
No good reason. Fixed.

MarkusH commented on django/db/migrations/state.py:
```diff
@@ -201,6 +200,7 @@ def register_model(self, app_label, model):
             self.app_configs[app_label].models = OrderedDict()
         self.app_configs[app_label].models[model._meta.model_name] = model
         self.clear_cache()
+        self.do_pending_operations(model)
```
As above.

MarkusH commented on django/db/models/fields/related.py:
```diff
@@ -27,74 +31,53 @@
+def add_lazy_relation(cls, field, relation, operation):
+    warnings.warn(
+        "add_lazy_relation() has been superseded by lazy_related_operation() "
+        "and related methods on the Apps class, and will be no longer be "
```
"Apps" --> "AppRegistry"

AlexHill replied:
Hmm? [django.apps.registry.Apps](https://github.com/django/django/blob/1379165b3548b843980e59201eea8961a0b30a2d/django/apps/registry.py#L12)

MarkusH commented on django/db/models/utils.py:
```diff
@@ -0,0 +1,27 @@
+    if isinstance(model, tuple):
+        return model
+    if isinstance(model, six.string_types):
+        return tuple(model.split("."))
```
This potentially returns a tuple with only one or more than 2 elements. I'd probably go with an approach as in https://github.com/django/django/pull/4097/files#diff-c8a3e1d248f8f5b10af7ac5b1d8479f4R26

AlexHill replied:
Yes, this could be stricter. I've now ensured it returns a 2-tuple, but have left out the handling for bare model names as that's taken care of in ``resolve_relation``.

AlexHill commented:
Updated with feedback from @aaugustin and @MarkusH. Notably:
* make_model_label is gone (we shouldn't need to go back to an "app_label.ModelName" string often)
* ``Apps.lazy_model_operation`` now only takes ``("app_label", "ModelName")`` tuples, not model instances or string labels. ``lazy_related_operation`` still works with all forms.

MarkusH commented:
buildbot, add to whitelist.

AlexHill commented:
See #4122 for details of the failing test.

carljm commented:
So this should be ready to go once #4122 is resolved to fix the failing test?

AlexHill commented:
@carljm yes, provided everyone's happy with the latest API changes.

carljm commented on django/db/models/fields/related.py:
```diff
@@ -27,74 +31,54 @@
+def lazy_related_operation(function, model, *related_models, **kwargs):
     """
-    Handle any pending relations to the sending model. Sent from class_prepared.
+    A wrapper for Apps.lazy_model_operation() where the first model has some
```
This docstring should perhaps also make note of the handling of `**kwargs`

AlexHill replied:
Good call. I've rewritten it.

carljm commented:
I like the latest API changes, and am happy with the state of this whenever the tests are passing.

aaugustin commented on django/apps/registry.py:
```diff
@@ -329,5 +333,43 @@ def clear_cache(self):
+        # If this function depends on more than one model, we recursively turn
+        # it into a chain of functions that accept a single model argument and
+        # pass each in turn to lazy_model_operation.
+        model_key, more_models = model_keys[0], model_keys[1:]
```
You could make the signature `lazy_model_operation(self, function, model_key, *more_model_keys)` (unless you consider that less obvious).

AlexHill replied:
I think I'd rather keep it as is. There's nothing special about the first argument to this method (unlike in ``lazy_related_operation``) and I'd prefer its signature to reflect that than to save a line here.

aaugustin commented on django/apps/registry.py:
```diff
@@ -45,8 +46,10 @@ def __init__(self, installed_apps=()):
 
-        # Pending lookups for lazy relations.
-        self._pending_lookups = {}
+        # Maps ("app_label", "ModelName") tuples to lists of functions to be
```
Currently the app registry uses "modelname" (`_meta.model_name`) rather than "ModelName" (`_meta.object_name`) everywhere because model names are treated case-insensitively. I don't think that's a very good design but for consistency I would prefer to keep using it.

AlexHill replied:
OK, I'll make ``make_model_tuple`` use ``_meta.model_name`` instead.

aaugustin commented on django/apps/registry.py:
```diff
@@ -329,5 +333,43 @@ def clear_cache(self):
 
+    def lazy_model_operation(self, function, *model_keys):
+        """
+        Take a function and a number of ("app_label", "ModelName") tuples, and
```
Same comment as above: "ModelName" => "modelname".

aaugustin commented on django/apps/registry.py:
```diff
@@ -329,5 +333,43 @@ def clear_cache(self):
+        try:
+            model_class = self.get_registered_model(*model_key)
+        except LookupError:
+            self._pending_operations[model_key].append(function)
```
Always for the same reason, you should call `.lower()` on `model_key[1]` before using it in `_pending_operations`.

aaugustin commented on django/apps/registry.py:
```diff
@@ -329,5 +333,43 @@ def clear_cache(self):
+        Take a newly-prepared model and pass it to each function waiting for
+        it. This is called at the very end of ``Apps.register_model()``.
+        """
+        key = model._meta.app_label, model._meta.object_name
```
And `model._meta.object_name` here.

aaugustin commented on django/apps/registry.py:
```diff
@@ -329,5 +333,43 @@ def clear_cache(self):
+        it. This is called at the very end of ``Apps.register_model()``.
+        """
+        key = model._meta.app_label, model._meta.object_name
+        for function in self._pending_operations.pop(key, []):
```
You can use `_pending_operations[key]` now that you made it a `defaultdict`.

AlexHill replied:
I kept ``pop`` here intentionally – I think it's cleaner than including a ``del`` statement.

aaugustin commented on docs/releases/1.9.txt:
```diff
@@ -260,6 +260,9 @@ Miscellaneous
   :class:`~django.forms.SelectDateWidget` in ``django.forms.widgets``
   (or simply ``django.forms``) instead.
 
+* ``django.db.models.fields.add_lazy_relation`` is deprecated. The new function
```
"Private API ``django.db..."

aaugustin commented on docs/releases/1.9.txt:
```diff
@@ -260,6 +260,9 @@ Miscellaneous
   (or simply ``django.forms``) instead.
 
+* ``django.db.models.fields.add_lazy_relation`` is deprecated. The new function
+  ``lazy_related_operation`` should be used instead.
```
Since it's a private API, I don't think we should mention it in the documentation. The warning message suffices.

aaugustin commented on tests/model_fields/tests.py:
```diff
@@ -211,13 +211,13 @@ def test_abstract_model_pending_lookups(self):
         fk_lookup = (AbstractForeignFieldsModel, opts.get_field('fk'))
         self.assertFalse(
-            any(lookup[0:2] == fk_lookup for lookup in opts.apps._pending_lookups.get(to_key, [])),
+            any(lookup[0:2] == fk_lookup for lookup in opts.apps._pending_operations.get(to_key, [])),
```
`pending_options[to_key]`

AlexHill replied:
In the interest of leaving _pending_operations untouched, could we leave these ``get`` calls as-is?

aaugustin commented on tests/model_fields/tests.py:
```diff
@@ -226,12 +226,12 @@ def test_abstract_model_pending_lookups(self):
         fk_lookup = (AbstractForeignFieldsModel, opts.get_field('m2m'))
         self.assertFalse(
-            any(lookup[0:2] == fk_lookup for lookup in opts.apps._pending_lookups.get(to_key, [])),
+            any(lookup[0:2] == fk_lookup for lookup in opts.apps._pending_operations.get(to_key, [])),
```
`pending_options[to_key]`

aaugustin commented on tests/model_fields/tests.py:
```diff
@@ -226,12 +226,12 @@ def test_abstract_model_pending_lookups(self):
         through_key = ('missing', 'Through')
         self.assertFalse(
-            any(lookup[0:2] == fk_lookup for lookup in opts.apps._pending_lookups.get(through_key, [])),
+            any(lookup[0:2] == fk_lookup for lookup in opts.apps._pending_operations.get(through_key, [])),
```
`pending_options[through_key]`

aaugustin commented:
I'm done commenting the changes to the app registry. The final implementation looks clean and consistent which is extremely important in this area. I read the rest of the patch and didn't spot anything suspicious but I'm a bit out of my depth there.

AlexHill commented:
I realised the tests Markus wrote no longer test the right thing after all my changes, so I've rewritten them and added a test to ``AppsTests`` to verify ``Apps.lazy_model_operation``.

AlexHill commented:
Ok, this is now just waiting for #4122 to be approved. Thanks for your help everyone!

AlexHill commented:
I've just rebased against current master to fix some merge conflicts, and rewritten the commit history to get rid of the junk: https://github.com/AlexHill/django/tree/lazy_ops_rebase

I'm keen to get this in so it doesn't stagnate further – if there's anything more I can do to help the process along please let me know!

aaugustin commented:
Jenkins says that the tests fail on SQLite. Can you have a look?

AlexHill commented:
@aaugustin the failures under SQLite are fixed by #4122, which I think is good to go as well.

AlexHill commented:
The squashed and rebased changes can be viewed in my [lazy_ops_rebase branch](https://github.com/AlexHill/django/commits/lazy_ops_rebase), up-to-date as of today. Should I squash the commits in this pull request or open a new one?

Secondly, for all tests to pass this depends on #4122, which has been approved in theory but needs final review.

Paging @akaariai and @carljm having expressed some interest in getting this committed :)

timgraham commented on django/apps/registry.py:
```diff
@@ -45,8 +46,10 @@ def __init__(self, installed_apps=()):
-        self._pending_lookups = {}
+        # Maps ("app_label", "modelname") tuples to lists of functions to be
+        # called when the corresponding model is ready. See this class'
+        # lazy_model_operation and do_pending_operations for details.
```
"methods for details"
(I also often write the names as `lazy_model_operation()` (with parenthesis) so it's obvious you're referring to a method and not an attribute.

timgraham commented on django/apps/registry.py:
```diff
@@ -332,5 +336,43 @@ def clear_cache(self):
+    def do_pending_operations(self, model):
+        """
+        Take a newly-prepared model and pass it to each function waiting for
+        it. This is called at the very end of ``Apps.register_model()``.
```
single backticks for consistency with elsewhere?

timgraham commented on tests/apps/tests.py:
```diff
@@ -259,6 +259,44 @@ def test_get_containing_app_config_apps_not_ready(self):
+        """
+        Tests apps.lazy_model_operation().
+        """
+
```
chop newline

timgraham commented on tests/model_fields/tests.py:
```diff
@@ -202,37 +202,44 @@ def test_related_name_converted_to_text(self):
+            class Meta:
+                abstract = True
+
+        assert AbstractForeignKeyModel._meta.apps is apps
```
self.assertIs

timgraham commented on tests/model_fields/tests.py:
```diff
@@ -202,37 +202,44 @@ def test_related_name_converted_to_text(self):
+                abstract = True
+
+        assert AbstractForeignKeyModel._meta.apps is apps
+        self.assertTrue(
```
assertEqual?

timgraham commented on django/apps/registry.py:
```diff
@@ -45,8 +46,10 @@ def __init__(self, installed_apps=()):
-        # Pending lookups for lazy relations.
-        self._pending_lookups = {}
+        # Maps ("app_label", "modelname") tuples to lists of functions to be
+        # called when the corresponding model is ready. See this class'
```
I think "class's" is preferred.

timgraham commented:
Let's wait for https://github.com/django/django/pull/4241 to merge first as that is a large patch and has some conflicts with this. I think that will happen in a few days.

AlexHill commented:
Thanks for your feedback Tim. I've followed your suggestions.

Don't mind waiting. Anssi mentioned that merging this patch could make his work on the rel stuff easier – is that still the case @akaariai?

akaariai commented:
This will likely make composite fields and related.py refactor easier later on.

AlexHill commented:
Alright, no need to merge this before #4241 then.

AlexHill commented:
Jenkins really doesn't seem to like some of these ``push --force``d commits. The build is failing on most configurations. I'll try closing and re-opening to re-trigger a build.

AlexHill commented:
OK @timgraham, I've rebased on top of master now that #4241 is in. There are a couple of those git errors in Jenkins but all tests are passing – I think this is good to go.

timgraham commented:
This seems like it's broken into more commits than needed. Also, none of them reference the ticket.

AlexHill commented:
Ok, I've put the bulk of this into a single commit referring to the ticket.

MarkusH commented on django/db/migrations/state.py:
```diff
@@ -213,23 +212,14 @@ def __init__(self, real_apps, models, ignore_swappable=False):
+        if ignore_swappable:
+            pending_models -= {make_model_tuple(settings.AUTH_USER_MODEL)}
+        if pending_models:
+            msg = "Unhandled pending operations for models: %s"
```
I actually like the error message raised before. This one doesn't really give me a hint on where to start debugging. Any chance you can get the field in that message again?

AlexHill replied:
Yeah, it is more informative – I initially catered for this by having ``Apps.lazy_model_operation`` take kwargs, but that bit of introspection was removed after some discussion in the previous PR: https://github.com/django/django/pull/3984#discussion-diff-24385523

It might be possible to get the field by introspecting the passed-in function object to see its partially-applied kwargs, but I think it would be a bit of fairly arcane code for the sake of this error message, and there are some usages which aren't related to a field at all. Would you be OK with this going in as-is for now and me having a crack at it in a followup?

MarkusH replied:
Yes, fine with me.

AlexHill replied:
Ok, I have this working, but it's arguably not worth the added complexity and needs some tests. I'll open a separate pull request to discuss it after this patch is committed.

timgraham commented:
merged in 720ff740e70e649a97fcf0232fec132569a52c7e and 7a75ed1a4fdf043064f0cce393bb93b698794978, thanks.

AlexHill commented:
Great! Thanks Tim and everyone else who took to the time to review this and offer feedback: @charettes, @MarkusH, @akaariai and particularly @carljm and @aaugustin. The patch that got committed is much, much nicer than what I initially proposed thanks to all of you.
